### PR TITLE
Fixes for most of valgrind errors related to unintialized values

### LINKED ISSUE #1999: P2P layer message padding data leak
**Bug Description**
The P2P layer pads message to a multiple of 16 bytes. This padding contains uninitialized bytes, which could contain security-relevant data.

**Impacts**
Describe which portion(s) of BitShares Core may be impacted by this bug. Please tick at least one box.
- [ ] API (the application programming interface)
- [ ] Build (the build process or something prior to compiled code)
- [ ] CLI (the command line wallet)
- [ ] Deployment (the deployment process after building such as Docker, Travis, etc.)
- [ ] DEX (the Decentralized EXchange, market engine, etc.)
- [x] P2P (the peer-to-peer network for transaction/block propagation)
- [ ] Performance (system or user efficiency, etc.)
- [ ] Protocol (the blockchain logic, consensus, validation, etc.)
- [x] Security (the security of system or user data, etc.)
- [ ] UX (the User Experience)
- [ ] Other (please add below)

**Expected Behavior**
The padding bytes should be initialized properly. Any "standard" padding scheme that does not change the number of blocks is acceptable, e. g. simple [zero-padding](https://en.wikipedia.org/wiki/Padding_(cryptography)#Zero_padding).

**Additional Context (optional)**
Thanks to @HarukaMa for noticing this.

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Bug Report
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

### PR DESCRIPTION
**Pull request by pmconrad**
Cherry-picked https://github.com/steemit/steem/commit/0bda58f55b69397e343de4c3587b2d035993fd94
Fixes #1999

### CODE REVIEW

File: libraries/net/node.cpp [hunk 1]
```diff
@@ -4965,18 +4965,33 @@ namespace graphene { namespace net { namespace detail {
 #  define INVOKE_AND_COLLECT_STATISTICS(method_name, ...) \
     try \
     { \
+<<<<<<< HEAD
```
Comment (abitmore) [inline libraries/net/node.cpp]: Please resolve these.

File: libraries/net/message_oriented_connection.cpp [hunk 2]
```diff
@@ -61,7 +63,7 @@ namespace graphene { namespace net {
       fc::time_point _last_message_sent_time;
 
       bool _send_message_in_progress;
-
+      std::atomic_bool readLoopInProgress;
```
Comment (abitmore) [inline libraries/net/message_oriented_connection.cpp]: Please change the variable name to be similar with others.

File: libraries/net/message_oriented_connection.cpp [hunk 3]
```diff
@@ -137,6 +140,20 @@ namespace graphene { namespace net {
...
     }
 
+    class THelper final
+    {
+      std::atomic_bool* Flag;
+    public:
+      explicit THelper(std::atomic_bool* flag) : Flag(flag)
+      {
+      FC_ASSERT(*flag == false, "Only one thread at time can visit it");
+      *flag = true;
```
Comment (abitmore) [inline libraries/net/message_oriented_connection.cpp]: Please update coding style. :)

File: libraries/net/message_oriented_connection.cpp [hunk 4]
```diff
@@ -137,6 +140,20 @@ namespace graphene { namespace net {
       _sock.bind(local_endpoint);
     }
 
+    class THelper final
```
Comment (abitmore) [inline libraries/net/message_oriented_connection.cpp]: Please update class name.

File: libraries/net/node.cpp (hunk 1 above)
Comment (pmconrad) [reply libraries/net/node.cpp]: oh, thanks. well, it *does* compile...

File: libraries/net/message_oriented_connection.cpp [hunk 5]
```diff
@@ -63,7 +63,7 @@ namespace graphene { namespace net {
       fc::time_point _last_message_sent_time;
 
       bool _send_message_in_progress;
```
Comment (abitmore) [inline libraries/net/message_oriented_connection.cpp]: While you're on it, perhaps update `_send_message_in_progress` to atomic and update the `assert` in the guarding code to `FC_ASSERT`?

Comment (pmconrad) [reply libraries/net/message_oriented_connection.cpp]: Replace `verify_no_send_in_progress` with `no_parallel_execution_guard`?

File: libraries/net/include/graphene/net/peer_connection.hpp [hunk 6]
```diff
@@ -258,17 +258,17 @@ namespace graphene { namespace net
...
+      firewall_check_state_data *firewall_check_state = nullptr;
 #ifndef NDEBUG
     private:
-      fc::thread* _thread;
-      unsigned _send_message_queue_tasks_running; // temporary debugging
+      fc::thread* _thread = nullptr;
+      unsigned _send_message_queue_tasks_running = 0; // temporary debugging
 #endif
-      bool _currently_handling_message; // true while we're in the middle of handling a message from the remote system
+      bool _currently_handling_message = false; // true while we're in the middle of handling a message from the remote system
```
Comment (abitmore) [inline libraries/net/include/graphene/net/peer_connection.hpp]: This is public in release build, but private in debug build. I think it's better to make it consistent.

File: libraries/net/message_oriented_connection.cpp (hunk 5 above)
Comment (abitmore) [reply libraries/net/message_oriented_connection.cpp]: Yes.

File: libraries/net/message_oriented_connection.cpp [hunk 7]
```diff
@@ -260,8 +279,12 @@ namespace graphene { namespace net {
         //pad the message we send to a multiple of 16 bytes
         size_t size_with_padding = 16 * ((size_of_message_and_header + 15) / 16);
         std::unique_ptr<char[]> padded_message(new char[size_with_padding]);
+
         memcpy(padded_message.get(), (char*)&message_to_send, sizeof(message_header));
         memcpy(padded_message.get() + sizeof(message_header), message_to_send.data.data(), message_to_send.size.value() );
+        char* paddingSpace = padded_message.get() + sizeof(message_header) + message_to_send.size.value();
```
Comment (abitmore) [inline libraries/net/message_oriented_connection.cpp]: Perhaps adapt coding style here too?

File: libraries/net/message_oriented_connection.cpp [hunk 8]
```diff
@@ -137,6 +140,20 @@ namespace graphene { namespace net {
...
+      std::atomic_bool* _flag;
+    public:
+      explicit no_parallel_execution_guard(std::atomic_bool* flag) : _flag(flag)
+      {
+         bool expected = false;
+         FC_ASSERT( flag->compare_exchange_strong( expected, true ), "Only one thread at time can visit it");
+      }
+      ~no_parallel_execution_guard()
+      {
+         *_flag = false;
```
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/net/message_oriented_connection.cpp]: `flag->compare_exchange_strong( a, b )` is the atomic equivalent of `if ( flag == a ) { flag = b; return true; } else return false;`. The second thread will throw an exception, but that happens in the constructor it means that the destructor is not executed, so it does not set the flag back to false. No locking needed.

File: libraries/net/message_oriented_connection.cpp (hunk 7 above)
Comment (pmconrad) [reply libraries/net/message_oriented_connection.cpp]: done

File: libraries/net/include/graphene/net/peer_connection.hpp (hunk 6 above)
Comment (pmconrad) [reply libraries/net/include/graphene/net/peer_connection.hpp]: done

File: libraries/net/message_oriented_connection.cpp (hunk 5 above)
Comment (pmconrad) [reply libraries/net/message_oriented_connection.cpp]: done

**Comment by pmconrad** [comment]
Rebased to avoid conflict

**Pull request merged**